### PR TITLE
[RISCV] Vectorize phi for loop carried @llvm.vector.reduce.fadd

### PR DESCRIPTION
**Pull request by lukel97**
LLVM vector reduction intrinsics return a scalar result, but on RISC-V vector reduction instructions write the result in the first element of a vector register. So when a reduction in a loop uses a scalar phi, we end up with unnecessary scalar moves:

```asm
loop:
    vfmv.s.f v10, fa0
    vfredosum.vs v8, v8, v10
    vfmv.f.s fa0, v8
```

This mainly affects ordered fadd reductions, since other types of reduction typically use element-wise vectorisation in the loop body. This tries to vectorize any scalar phis that feed into a fadd reduction in RISCVCodeGenPrepare, converting:

```llvm
loop:
%phi = phi <float> [ ..., %entry ], [ %acc, %loop]
%acc = call float @llvm.vector.reduce.fadd.nxv4f32(float %phi, <vscale x 2 x float> %vec)
```

to

```llvm
loop:
%phi = phi <vscale x 2 x float> [ ..., %entry ], [ %acc.vec, %loop]
%phi.scalar = extractelement <vscale x 2 x float> %phi, i64 0
%acc = call float @llvm.vector.reduce.fadd.nxv4f32(float %x, <vscale x 2 x float> %vec)
%acc.vec = insertelement <vscale x 2 x float> poison, float %acc.next, i64 0
```

Which eliminates the scalar -> vector -> scalar crossing during instruction selection.

> [!NOTE]
> The tests were added in a separate commit in this PR, in case they needed reviewed too. The codegen diff should be viewable between the two commits.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/riscv-codegenprepare-asm.ll [hunk 1]
```diff
@@ -44,21 +44,21 @@ define float @reduce_fadd(ptr nocapture noundef readonly %f, i32 noundef signext
...
 ; CHECK-NEXT:    mv a4, s2
 ; CHECK-NEXT:    mv a5, a2
 ; CHECK-NEXT:  .LBB0_5: # %vector.body
 ; CHECK-NEXT:    # =>This Inner Loop Header: Depth=1
-; CHECK-NEXT:    vl2re32.v v8, (a4)
-; CHECK-NEXT:    vfmv.s.f v10, fa0
-; CHECK-NEXT:    vfredosum.vs v8, v8, v10
-; CHECK-NEXT:    vfmv.f.s fa0, v8
+; CHECK-NEXT:    vl2re32.v v10, (a4)
+; CHECK-NEXT:    vsetvli a6, zero, e32, m2, ta, ma
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/riscv-codegenprepare-asm.ll:55]: This looks like a gap in RISCVInsertVSETVLI. Previously we were doing PRE, but now that the `vmv.s.x` was moved to the loop preheader it no longer kicks in. The M2 vsetvli could subsume the vsetvl in the preheader entirely.

This happens with and without 286a366d057112a112507b9e34c68d35b3b820f7

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp [hunk 2]
```diff
@@ -103,6 +105,61 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
...
+// %acc.next, i64 0
+//
+// Which eliminates the scalar -> vector -> scalar crossing during instruction
+// selection.
+bool RISCVCodeGenPrepare::visitIntrinsicInst(IntrinsicInst &I) {
+  if (I.getIntrinsicID() != Intrinsic::vector_reduce_fadd)
+    return false;
+
+  auto *PHI = dyn_cast<PHINode>(I.getOperand(0));
+  if (!PHI || !llvm::is_contained(PHI->incoming_values(), &I))
```
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:143]: Sure, we could try to generalise it to all reduction intrinsics then. We would need to double check that wrapping every reduction in extract_element/insert_element doesn't negatively affect codegen in other places though.

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp [hunk 3]
```diff
@@ -103,6 +105,61 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
...
+// float> %vec)
+//
+// ->
+//
+// loop:
+// %phi = phi <vscale x 2 x float> [ ..., %entry ], [ %acc.vec, %loop]
+// %phi.scalar = extractelement <vscale x 2 x float> %phi, i64 0
+// %acc = call float @llvm.vector.reduce.fadd.nxv4f32(float %x, <vscale x 2 x
+// float> %vec) %acc.vec = insertelement <vscale x 2 x float> poison, float
+// %acc.next, i64 0
```
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:134]: Woops, clang-format doesn't seem to like this.

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:143]: Actually ignore the above, I think llvm.vector.reduce.fadd is the only intrinsic applicable here: We expand llvm.vector.reduce.fmul and llvm.vector.reduce.fmax/fmin don't have a no-fast-math accumulator argument

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp [hunk 4]
```diff
@@ -157,6 +155,9 @@ bool RISCVCodeGenPrepare::visitIntrinsicInst(IntrinsicInst &I) {
   Builder.SetInsertPoint(&I);
   I.setOperand(0, Builder.CreateExtractElement(VecPHI, (uint64_t)0));
 
+  if (PHI->hasNUses(0))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:158]: use_empty()?

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp [hunk 5]
```diff
@@ -103,6 +105,62 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
...
+  if (!PHI)
+    return false;
+
+  Type *VecTy = I.getOperand(1)->getType();
+  IRBuilder<> Builder(PHI);
+  auto *VecPHI = Builder.CreatePHI(VecTy, PHI->getNumIncomingValues());
+
+  for (auto *BB : PHI->blocks()) {
+    Builder.SetInsertPoint(BB->getTerminator());
+    Value *InsertElt = Builder.CreateInsertElement(
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:150]: Doesn't look like we verified that we're in a loop. If the phi is just the merge of ocntrol flow from an if/else or something does this unnecessarily hoist an insertelement into the if/else?

Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:150]: The initial revision of this PR checked that the reduction was one of the incoming values in phi: 05488f67033b816182134716aff97a0fce9939b4

but I relaxed it to address the comment in https://github.com/llvm/llvm-project/pull/78244#discussion_r1453391718

I haven't included any tests for non-loop control flow though. @dtcxzyw should we maybe start with this patch restricted to loops and then relax it in a follow up patch?

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:150]: Please revert to the loop specific version.  Please also restrict this - in the initial patch - to the case where the only use of the phi is the intrinsic.  Let's start with something vary narrowly focused on the motivating case and generalize in separate patches if desired.

Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:150]: > I haven't included any tests for non-loop control flow though. @dtcxzyw should we maybe start with this patch restricted to loops and then relax it in a follow up patch?

Yeah, we can generalize it in the future if we find it exists in some real-world applications.

File: llvm/test/CodeGen/RISCV/rvv/riscv-codegenprepare-asm.ll [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+; CHECK-NEXT:  .LBB0_9:
+; CHECK-NEXT:    ld ra, 40(sp) # 8-byte Folded Reload
+; CHECK-NEXT:    ld s0, 32(sp) # 8-byte Folded Reload
+; CHECK-NEXT:    ld s1, 24(sp) # 8-byte Folded Reload
+; CHECK-NEXT:    ld s2, 16(sp) # 8-byte Folded Reload
+; CHECK-NEXT:    ld s3, 8(sp) # 8-byte Folded Reload
+; CHECK-NEXT:    addi sp, sp, 48
+; CHECK-NEXT:    ret
+entry:
+  %cmp4 = icmp sgt i32 %N, 0
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/riscv-codegenprepare-asm.ll:83]: For both of these tests, you don't need the whole loop structure emitted by loop vectorizer.  You just need the main vector loop.

**Review by preames — APPROVED** [review]
LGTM w/minor comment addressed.

File: llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp [hunk 7]
```diff
@@ -103,6 +105,62 @@ bool RISCVCodeGenPrepare::visitAnd(BinaryOperator &BO) {
...
+//
+// Which eliminates the scalar -> vector -> scalar crossing during instruction
+// selection.
+bool RISCVCodeGenPrepare::visitIntrinsicInst(IntrinsicInst &I) {
+  if (I.getIntrinsicID() != Intrinsic::vector_reduce_fadd)
+    return false;
+
+  auto *PHI = dyn_cast<PHINode>(I.getOperand(0));
+  if (!PHI || !PHI->hasOneUse() ||
+      !llvm::is_contained(PHI->incoming_values(), &I))
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp:142]: Just for the record... This isn't checking that the phi is in a loop.  It's checking it is in a cycle, which is not quite the same thing.  In particular, we allow both non-loop cycles, and loops in unreachable code which aren't "loops".  Not a problem here, just pointing it out for future reference.

**Pull request merged**